### PR TITLE
Add ifThenElse constructor for HttpMiddleware

### PR DESCRIPTION
**Pull request by pravin-raha**
See #454

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala [hunk 1]
```diff
@@ -52,6 +52,23 @@ object HttpMiddleware {
 
   type RequestP[+A] = (Method, URL, List[Header]) => A
 
+  def ifThenElse[R, E](cond: RequestP[Boolean])(left: HttpApp[R, E], right: HttpApp[R, E]): HttpMiddleware[R, E] =
+    HttpMiddleware.fromMiddlewareFunction((m, u, h) =>
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala]: We can add a fromHttpFunction for simplicity.

File: zio-http/src/test/scala/zhttp/middleware/HttpMiddlewareSpec.scala [hunk 2]
```diff
@@ -14,6 +14,14 @@ object HttpMiddlewareSpec extends DefaultRunnableSpec {
     UIO(Response.ok).delay(1 second)
   }
 
+  val app1: HttpApp[Any with Clock, Nothing] = HttpApp.collectM { case Method.GET -> !! / "create" =>
+    UIO(Response.status(Status.CREATED)).delay(1 second)
+  }
+
+  val app2: HttpApp[Any with Clock, Nothing] = HttpApp.collectM { case Method.GET -> !! / "bad_request" =>
+    UIO(Response.status(Status.BAD_REQUEST)).delay(1 second)
+  }
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/middleware/HttpMiddlewareSpec.scala:23]: We can use simpler apps for testing such as 

```scala
HttpApp.ok.addHeader("name", "left")
```

Then assert using "hasHeader"

File: zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala (hunk 1 above)
Comment (pravin-raha) [reply zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala]: I don't understand. what you want. What will be the  signature of `fromHttpFunction`

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala]:
```scala
def fromHttpFunction[R, E](f: RequestP[HttpApp[R, E]]): HttpMiddleware[R, E] = ???
def fromHttpFunctionM[R, E](f: RequestP[ZIO[R, E, HttpApp[R, E]]]): HttpMiddleware[R, E] = ???
```

Comment (pravin-raha) [reply zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala]: Added those new methods

File: zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala [hunk 3]
```diff
@@ -52,6 +52,17 @@ object HttpMiddleware {
 
   type RequestP[+A] = (Method, URL, List[Header]) => A
 
+  def ifThenElse[R, E](cond: RequestP[Boolean])(left: HttpApp[R, E], right: HttpApp[R, E]): HttpMiddleware[R, E] =
+    HttpMiddleware.fromHttpFunction((m, u, h) => if (cond(m, u, h)) left else right)
+
+  def ifThenElseM[R, E](cond: RequestP[UIO[Boolean]])(left: HttpApp[R, E], right: HttpApp[R, E]): HttpMiddleware[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala:58]:
```suggestion
  def ifThenElseM[R, E](cond: RequestP[ZIO[R, E, Boolean]])(left: HttpApp[R, E], right: HttpApp[R, E]): HttpMiddleware[R, E] =
```

Comment (pravin-raha) [reply zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala:58]: can we go with `cond: RequestP[ZIO[R, Option[E], Boolean]]` we are using `Option[E]` every were, like 
`def modifyM[R1 <: R, E1 >: E](f: RequestP[ZIO[R1, Option[E1], HttpMiddleware[R1, E1]]]): HttpMiddleware[R1, E1]`

**Comment by dinojohn** [comment]
Cherry-picked the commits from this branch onto #485.

**Comment by dinojohn** [comment]
closing in favour of #485

**Pull request closed without merging**